### ISSUE TITLE
Support playback of videos with multiple versions

### ISSUE THREAD

**Issue by Montellese** [Montellese/mediaimporter.plex#29]
Plex seems to support multiple versions for a video (movie / episode). We need to find a way to support this in `mediaimporter.plex`.

**Comment by Montellese** [comment]
I tried it with version `0.0.9` and for me it simply plays the "best" version of a movie but playback works fine.

**Comment by rijsab** [comment]
Is there anything different between 0.0.9 and 0.0.10? I have tried all files with multiple versions and none of them played.. only 1 version of a file was able to play, the other version gave playback error. So for example if I have File X with version A & B, only version B was able to play, version A is giving playback error.

**Comment by Montellese** [comment]
I tested with `0.0.9` and with `improvement/retry_requests` (`pre-0.0.10`)  and it worked fine. But there are probably differences between our setups.

How were you able to choose between version A and B of File X since `mediaimporter.plex` doesn't support this yet?

**Comment by rijsab** [comment]
When I enter the library on Kodi I can see 2 versions sitting next to each other.. 2 posters of same movie. One is playing, the other not. Would be good if I can choose which version to play. For example if I already played version A on my plex, this version will show up as continue to watch. However with current setup I cannot play this on Kodi as only version B is working and media importer selects automatically version A. So I have to look for version B in the library first..

**Comment by rijsab** [comment]
Hi Montellese,

It would be really great to have a pop up for selecting which file version to play, preferably in plex media importer 0.0.10.

Is this something we can expect soon?

Also curious why 0.0.9 is showing as latest, but 0.0.10 seems working fine as well but also for larger libraries?

**Comment by Montellese** [comment]
> Hi Montellese,
> 
> It would be really great to have a pop up for selecting which file version to play, preferably in plex media importer 0.0.10.

I will probably not be able to provide this as part of Kodi's normal play behaviour but I can probably provide a context menu item to explicitely choose between different versions (similar to Plex Web). 

> Is this something we can expect soon?

I can't make any promises since I have other important issues to fix in the general media import logic.

> Also curious why 0.0.9 is showing as latest, but 0.0.10 seems working fine as well but also for larger libraries?

0.0.9 is the last official release. For 0.0.10 I have already done some improvements which haven't been officially released. The parallel import improvement and the retry logic are still experimental since I can't test everything myself which is why I'm waiting on feedback from users (like you). But I plan to release these changes soon.

**Comment by Montellese** [comment]
> When I enter the library on Kodi I can see 2 versions sitting next to each other.. 2 posters of same movie. One is playing, the other not. Would be good if I can choose which version to play. For example if I already played version A on my plex, this version will show up as continue to watch. However with current setup I cannot play this on Kodi as only version B is working and media importer selects automatically version A. So I have to look for version B in the library first..

That's odd. I faked a multi version movie and I still only see one movie in Kodi and mediaimporter.plex automatically chose the "first" version.

**Comment by rijsab** [comment]
I believe this needs to be turned on through Kodi Settings. I definitely can see 2 versions of some media instead of 1 in my library. I have to see if I can upload a picture of this. I know PlexKodiConnect can let you choose between versions of 1 file when I hit play.. however I find media importer much better

**Comment by Montellese** [comment]
> I know PlexKodiConnect can let you choose between versions of 1 file when I hit play.. however I find media importer much better 

Yes it is possible but it would require me to change a lot of stuff in mediaimporter. plex and it would probably slow down playback start.

I'm sure that someone will add this functionality in the future but right now it's not high priority.

**Comment by rijsab** [comment]
Unfortunately pal, I hope someone can support on different versions of files as I really want to keep media importer. 

Did you try to have different versions of a file and were they playing correctly? Not sure how things works, but if one version fails, is there an attempt to play the second file?

**Comment by Montellese** [comment]
> Did you try to have different versions of a file and were they playing correctly? Not sure how things works, but if one version fails, is there an attempt to play the second file?

As mentioned before I added a second version of an existing movie but `mediaimporter.plex` only imported one of them and that one played fine. I can't test if it would also properly play the other version because it isn't available in my imported library.

**Comment by rijsab** [comment]
Thanks - quiet strange as I have imported different versions of a movie and not just one.. do you want me to upload a log?

**Comment by Montellese** [comment]
A log probably won't help because it doesn't contain information about specific media items as long as there was no issue during the import / synchronization process. But it would be great if you could provide a screenshot from Plex Web with the multiple versions and one from the imported library in Kodi with the multiple movies for the different versions.

**Comment by rijsab** [comment]
OK, so I have noticed the following. I have a 4k library and a movies library. For example.. I have 2 fast 2 furious available in 4k folder but also in 1080p movies folder. Here is were I believe the issue starts.. the movies library file is having a 1080p 35bit & 1080p 15bit. Plex is showing the 4k movie file and movies 1080p file side by side. However the 35bit and 15 bit can only been seen by entering the 1080p file in movies folder. If only 4k file is available together with just one version of 1080p all seems to work..

**Comment by Montellese** [comment]
OK so you have three versions of the same movie in two differente Plex libraries. One is 4k without an alternative version and the other two are 1080p with two different versions which Plex Web shows as a single movie with the "Play Version..." possibility. Is that correct? And I assume you can play the 4k version fine but the 1080p version doesn't work?

**Comment by rijsab** [comment]
All correct mate Plex Web shows the 1080 as one poster, when selecting I can choose which version to play, otherwise Plex picks the best resolution depending on the platform being used.. 4K as yih have mentioned plays fine because there is only 1 version available for it.

**Comment by rijsab** [comment]
Hi Montellese,

I really want to urge you if this function can be implemented and prioritized please. Plex media importer is currently useless for me due to this missing functionality. Both PlexKodiConnect as Composite are having this, however I want to keep media importer due to its speed. 

Apologies for bothering you again on this, but I hope you understand were I am coming from and you already gave me the full support. 

I have made a feature request at LibreElec to implement media importer in their releases as well.

Many thanks for considering this!

**Comment by Montellese** [comment]
Can you give https://github.com/Montellese/mediaimporter.plex/archive/refs/heads/feature/context_menu.zip a try? This branch contains two improvements:
* I changed the logic to determine the stream URL a bit so it considers more than just the first possible stream in case it can't create a proper streaming URL for it. But I don't think that it will fix your problem.
* I added context menu integration for media items imported with `mediaimporter.plex`. Once the media provider has been synced by Kodi after startup you can open the context menu on any media item imported with `mediaimporter.plex` choose `Plex Media Importer` and then choose `Play ...`. Then you can either choose to play the default stream or if the media item has multiple versions you can choose `Play Version ...` and choose between the different versions.

Please let me know if the context menu integration works for you and whether you can play both versions of your media items using the new `Play Version ...` or not.

**Comment by rijsab** [comment]
Hi Montellese, 

Thank you for putting your efforts in this!

I have tried the context menu, however when I click on Play version nothing happens?

When I Play directly same issue persists with Play error showing up. 

Thanks.

**Comment by rijsab** [comment]
Logfile after receiving Plex Media importer error when I hit the context menu.

**Comment by Montellese** [comment]
Thanks for the logs. That's my bad. It should be fixed in the latest version of the branch so just download the ZIP again from https://github.com/Montellese/mediaimporter.plex/archive/refs/heads/feature/context_menu.zip and try it again.

Concerning the playback failure could you also look into the logs of the Plex server to see why it returns HTTP 503 for the playback request? To get the playback URL of a media item I ask `python-plexapi` for a URL for the specific item and that's what `mediaimporter.plex` uses everywhere. So I don't know why it should fail for media items with multiple versions and work for the rest. With my server everything works fine.

It would also be interesting to see what PKC uses as playback URL for these items.

**Comment by rijsab** [comment]
Hi Montellese,

I have requested the owner of the server and hope to come back to you soon on this. 

In the mean time I have tried the new zip file and now I was able to select the version, however still getting a play back error. I have played the version with the highest bitrate on Plex web server and everything plays fine, so no issue at server side. 

Please see logfile..

http://ix.io/3qEl

I feel we are almost there now I am able to select which version bit not sure why still play back error..

**Comment by rijsab** [comment]
Hi,

All working now! The server owner did some clean up on urls. I did a fresh install and until now all working, also when hitting play without using the context menu!

Really appreciate the quick solutions for this.

Will keep an eye on your work!

Take care.

**Comment by Montellese** [comment]
Thanks for the feedback. I'll close this issue once I've merged the `feature/context_menu` branch into `master`.
